Unify invoking JIT functions

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* #57856 Unify assignment of OwnerRRef result
* #57855 Make processPythonExecution return a future
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* **#57851 Unify invoking JIT functions**
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

The same as the previous PR, but for JIT functions.

Differential Revision: [D28253841](https://our.internmc.facebook.com/intern/diff/D28253841/)

lw committed:
Unify invoking JIT functions

The same as the previous PR, but for JIT functions.

Differential Revision: [D28253841](https://our.internmc.facebook.com/intern/diff/D28253841/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT functions"


The same as the previous PR, but for JIT functions.

Differential Revision: [D28253841](https://our.internmc.facebook.com/intern/diff/D28253841/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT functions"


The same as the previous PR, but for JIT functions.

Differential Revision: [D28253841](https://our.internmc.facebook.com/intern/diff/D28253841/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/script_call.h:
```diff
@@ -31,6 +31,7 @@ class TORCH_API ScriptCall : public RpcCommandBase {
   bool hasOp() const;
   std::shared_ptr<Operator> op() const;
   bool hasQualifiedName() const;
+  // FIXME It's weird to return a const _value_, should this return a reference?
```
Yep, returning const reference look better to me too.

lw replied:
I'll change it then

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.h:
```diff
@@ -56,6 +56,12 @@ class TORCH_API RequestCallbackImpl : public RequestCallbackNoPython {
+
+  // Helpers to run user-defined functions, operators and other computations.
+
+  c10::intrusive_ptr<JitFuture> runJitFunction(
```
similar as the comment in the previous PR, this does not need to be a member function?

lw replied:
It will be used by the subclass later in the stack

mrshenli replied:
Oh I see, we are adding a new subclass of `RequestCallbackImpl`? IIUC, it does not has a subclass today?

lw replied:
Oops, no, got confused with the previous PR. This method in fact will just be used in here, so it could perhaps be a simple function, though I think it might use some other methods in turn, and also making a method is more "consistent" with `runJitOperator`. Does this make sense?

mrshenli replied:
Yep, LGTM

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -282,23 +269,9 @@ void RequestCallbackImpl::processScriptRemoteCall(
-                    .jitCompilationUnit()
-                    ->get_function(scriptRemoteCall.qualifiedName())
-                    .runAsync(stack);
-    if (jitFuture->completed()) { // short-cut.
```
we are skipping fast passes, let's measure the impact.

lw replied:
What benchmarks would you like me to run?

Also, I don't remember if I said it clearly before, so let me explain why I think it's better to avoid these fastpaths/shortcuts: they are terribly dangerous for CUDA sync, because at no point we're synchronizing the events contained in the future with the current streams at the callsite. (It can be done, one just has to call `future->wait()`, but we didn't do it anywhere, and it's super easy to forget about).

Moreover, I believe they hurt readability, because they force us to lay out the code in a different order than how it's executed. Namely, now we need to do something like this:
```
continuation = []() { bar(); };
auto fut = foo();
if (fut->completed()) {
  continuation();
} else {
  fut->addCallback(continuation);
}
```
Note how in the code `bar()` appears _before_ `foo()`, whereas in the "logical flow" of the program `bar()` comes _after_ `foo()`. This requires the reader to "jump around" and makes it harder to quickly understand what the code is doing. Without shortcuts, the above code becomes:
```
auto fut = foo();
fut->addCallback([]() { bar(); });
```

As for the supposed performance benefit of these fastpaths, I honestly don't think that we're introducing a regression. The `addCallback` method itself already has such a shortcut logic internally, so in fact these fastpaths were redundant. Moreover, after the changes in #57638, invoking `addCallback` should come with literally zero overhead, and the compiler will most likely inline it all the time. (I said `addCallback` but I also mean `then` since they're basically the same).

mrshenli replied:
IIRC, @jjlilley added this a while ago when optimizing PyPer perf. So, I assume there was a reason to introduce fast path at least at that time. 

> Moreover, after the changes in #57638, invoking addCallback should come with literally zero overhead, and the compiler will most likely inline it all the time. 

If we are confident there is no additional overhead introduce, this looks good to me.

mrshenli reviewed (approved):
LGTM

lw committed:
Update on "Unify invoking JIT functions"


The same as the previous PR, but for JIT functions.

Differential Revision: [D28253841](https://our.internmc.facebook.com/intern/diff/D28253841/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT functions"


The same as the previous PR, but for JIT functions.

Differential Revision: [D28253841](https://our.internmc.facebook.com/intern/diff/D28253841/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT functions"


The same as the previous PR, but for JIT functions.

Differential Revision: [D28253841](https://our.internmc.facebook.com/intern/diff/D28253841/)

[ghstack-poisoned]

lw committed:
Update on "Unify invoking JIT functions"


The same as the previous PR, but for JIT functions.

Differential Revision: [D28253841](https://our.internmc.facebook.com/intern/diff/D28253841/)

[ghstack-poisoned]
